### ISSUE TITLE
Android client can't connect if many messages in the queue

### ISSUE THREAD

**Issue by vernazza** [Azure/azure-iot-sdk-java#334]
- **OS and version used:** <VERSION> Android

- **Java runtime used:** <VERSION> Android SDK 28

- **SDK version used:** <VERSION> com.microsoft.azure.sdk.iot:iot-device-client:1.14.0

# Description of the issue:
When in the Iot message queue there are many messages in the cloud to device message queue (i.e. at least 12), the iothub client can't connect.
The `DeviceClient.open()` function throws an exception, after very few mseconds

# Code sample exhibiting the issue:
>                deviceClient = new DeviceClient(connectionString, IotHubClientProtocol.MQTT);
>                deviceClient.setMessageCallback(this, this);
>                deviceClient.registerConnectionStatusChangeCallback(this, this);
>                deviceClient.open();

# Console log of the issue:
An exception is thrown. The exception is
`java.io.IOException: Could not open the connection`

The cause field of the exception is:
`com.microsoft.azure.sdk.iot.device.exceptions.ProtocolException: Unable to subscribe to topic :devices/DEVICEID/messages/devicebound/#`

And the cause field in the cause field is
`Timed out waiting for a response from the server (32000)`

The exception is thrown very quickly, so it doesn't seem to match the "timeout" reason

**Comment by timtay-microsoft** [comment]
Are you seeing this behavior consistently? And what happens if you use AMQPS?

**Comment by vernazza** [comment]
I did some extra tests and:
- yes, the behaviour is consistent
- the problem appears when there are 12 or more messages. It doesn't matter the size of the message. 11 messages are ok, 12 or more aren't
- the problem is happening with MQTT and MQTT_WS; while using AMQPS or AMQPS_WS the library works fine

**Comment by jasmineymlo** [comment]
@vernazza, sorry for the delay.  I am able to repro this bug and it's not only on Android.  Investigating and will keep you posted.  Thanks.

**Comment by timwieman** [comment]
+1. Seeing this on Linux Ubuntu 18.04

**Comment by timwieman** [comment]
This is the stack trace for the exception.  Note that when stepping through in the debugger in VS Code, the connect exception is *not* hit. 
```
java.io.IOException: Could not open the connection
    at com.microsoft.azure.sdk.iot.device.DeviceIO.open (DeviceIO.java:164)
    at com.microsoft.azure.sdk.iot.device.InternalClient.open (InternalClient.java:130)
    at com.microsoft.azure.sdk.iot.device.DeviceClient.open (DeviceClient.java:317)
    at com.microsoft.docs.iothub.samples.SimulatedDevice.main (SimulatedDevice.java:61)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
    at java.lang.Thread.run (Thread.java:844)
Caused by: com.microsoft.azure.sdk.iot.device.exceptions.ProtocolException: Unable to subscribe to topic :devices/hackathonDevice01/messages/devicebound/#
    at com.microsoft.azure.sdk.iot.device.transport.mqtt.exceptions.PahoExceptionTranslator.convertToMqttException (PahoExceptionTranslator.java:75)
    at com.microsoft.azure.sdk.iot.device.transport.mqtt.Mqtt.subscribe (Mqtt.java:283)
    at com.microsoft.azure.sdk.iot.device.transport.mqtt.MqttMessaging.start (MqttMessaging.java:61)
    at com.microsoft.azure.sdk.iot.device.transport.mqtt.MqttIotHubConnection.open (MqttIotHubConnection.java:179)
    at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.openConnection (IotHubTransport.java:640)
    at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.open (IotHubTransport.java:277)
    at com.microsoft.azure.sdk.iot.device.DeviceIO.open (DeviceIO.java:160)
    at com.microsoft.azure.sdk.iot.device.InternalClient.open (InternalClient.java:130)
    at com.microsoft.azure.sdk.iot.device.DeviceClient.open (DeviceClient.java:317)
    at com.microsoft.docs.iothub.samples.SimulatedDevice.main (SimulatedDevice.java:61)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
    at java.lang.Thread.run (Thread.java:844)
Caused by: org.eclipse.paho.client.mqttv3.MqttException: Timed out waiting for a response from the server
    at org.eclipse.paho.client.mqttv3.internal.Token.waitForCompletion (Token.java:112)
    at org.eclipse.paho.client.mqttv3.MqttToken.waitForCompletion (MqttToken.java:67)
    at com.microsoft.azure.sdk.iot.device.transport.mqtt.Mqtt.subscribe (Mqtt.java:278)
    at com.microsoft.azure.sdk.iot.device.transport.mqtt.MqttMessaging.start (MqttMessaging.java:61)
    at com.microsoft.azure.sdk.iot.device.transport.mqtt.MqttIotHubConnection.open (MqttIotHubConnection.java:179)
    at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.openConnection (IotHubTransport.java:640)
    at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.open (IotHubTransport.java:277)
    at com.microsoft.azure.sdk.iot.device.DeviceIO.open (DeviceIO.java:160)
    at com.microsoft.azure.sdk.iot.device.InternalClient.open (InternalClient.java:130)
    at com.microsoft.azure.sdk.iot.device.DeviceClient.open (DeviceClient.java:317)
    at com.microsoft.docs.iothub.samples.SimulatedDevice.main (SimulatedDevice.java:61)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
    at java.lang.Thread.run (Thread.java:844)
```

**Comment by davilu** [comment]
Working on it. It will be my priority.

**Comment by az-iot-builder-01** [comment]
@vernazza, @timwieman, @AiyappaND, @davilu, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by yzhong94** [comment]
@AiyappaND @timwieman @vernazza Let us know if the referenced PR fixed the issues.

**Comment by jorgeperezruiz** [comment]
Hi, I am using `1.15.0` and I am still getting this issue when trying to open the connection. 

`com.microsoft.azure.sdk.iot.device.exceptions.ProtocolException: Unable to subscribe to topic :devices/<DEVICE_ID>/messages/devicebound/#`

The protocol we are using is `IotHubClientProtocol.MQTT`

**Comment by davilu** [comment]
@jorgeperezruiz There is no `1.15.0` release for `Microsoft.Azure.Devices.Client`. The release I was referring to is `1.19.0` with `Microsoft.Azure.Devices.Shared` release `1.15.2`. Please let me know how it goes.

**Comment by jorgeperezruiz** [comment]
@davilu I am talking about https://github.com/Azure/azure-iot-sdk-java/releases/tag/LTS_01_2019_Rev01

`Java IotHub Device Client (com.microsoft.azure.sdk.iot:iot-device-client:1.15.0)`

**Comment by jasmineymlo** [comment]
@jorgeperezruiz cef10d279ac744045094d7bffb98e1455d9095ec is the fix for the issue and the next release will contain the fix.  Thanks.

**Comment by jorgeperezruiz** [comment]
@jasmineymlo Hi, we are seeing the same issue also happening with `AMQPS`, is that expected? However, it is not happening with `HTTPS`. 

Do we have an ETA for the next release?

Thanks a lot.

**Comment by timtay-microsoft** [comment]
The fix for this issue has been released as of Device Client 1.16.0. Please try out the new version and let us know if you are still seeing this issue. Since the fix has been released, I'm going to close this issue now.
